### ISSUE TITLE
Can't pull strimzi/operator when version specified

### ISSUE THREAD

**Issue by TFBW** [rancher/rke2#1234]
**Environmental Info:**
RKE2 Version: 
rke2 version v1.21.2+rke2r1 (d58ad6197b05144711671cec03a9027058969bf0)
go version go1.16.4b7

Node(s) CPU architecture, OS, and Version: 
Linux hcm06 3.10.0-1160.31.1.el7.x86_64 #1 SMP Thu Jun 10 13:32:12 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:
3 servers on VMs in an isolated subnet (evaluation)

**Describe the bug:**

I'm evaluating RKE2 on VMs running in isolation: they have no Internet access except via HTTP proxy and a docker pull-through cache registry service. I have configured /etc/rancher/rke2/registries.yaml accordingly, and things are generally working. For my next trick, I wanted to deploy strimzi/operator, and have hit an odd snag: the deployment keeps failing to pull the image. I flailed around with the registries.yaml a bit, but am now convinced that this is an RKE2 problem, or at least a problem with the containerd that it bundles.

**Steps To Reproduce:**

- Configured environment with registry mirror and HTTP proxies
- Installed RKE2
- Followed https://strimzi.io/docs/operators/latest/quickstart.html

**Expected behavior:**

Expected operator deployment to complete successfully.

**Actual behavior:**

Deployment does not complete: repeatedly fails to fetch image "quay.io/strimzi/operator:0.24.0".

**Additional context / logs:**

I was able to pull the image using `docker` on the host which is providing the registry mirror, but not using `crictl` on one of my RKE2 servers. However, if I remove the version tag, then `crictl` *can* pull the "latest" image. The following output demonstrates this, with debug detail.

```
[root@hcm06 ~]# crictl pull quay.io/strimzi/operator:0.24.0
DEBU[0000] get image connection                         
DEBU[0000] connect using endpoint 'unix:///run/k3s/containerd/containerd.sock' with '5s' timeout 
DEBU[0000] connected successfully using endpoint: unix:///run/k3s/containerd/containerd.sock 
DEBU[0000] PullImageRequest: &PullImageRequest{Image:&ImageSpec{Image:quay.io/strimzi/operator:0.24.0,Annotations:map[string]string{},},Auth:nil,SandboxConfig:nil,} 
DEBU[0002] PullImageResponse: nil                       
FATA[0002] pulling image: rpc error: code = Unknown desc = failed to pull and unpack image "quay.io/strimzi/operator:0.24.0": failed to resolve reference "quay.io/strimzi/operator:0.24.0": failed to do request: Head "https://quay.io/v2/strimzi/operator/manifests/0.24.0": dial tcp: lookup quay.io on 192.168.204.1:53: server misbehaving 
[root@hcm06 ~]# crictl pull quay.io/strimzi/operator
DEBU[0000] get image connection                         
DEBU[0000] connect using endpoint 'unix:///run/k3s/containerd/containerd.sock' with '5s' timeout 
DEBU[0000] connected successfully using endpoint: unix:///run/k3s/containerd/containerd.sock 
DEBU[0000] PullImageRequest: &PullImageRequest{Image:&ImageSpec{Image:quay.io/strimzi/operator,Annotations:map[string]string{},},Auth:nil,SandboxConfig:nil,} 
DEBU[0002] PullImageResponse: &PullImageResponse{ImageRef:sha256:b360b154b122e3bb16ca12c70517ba0dce04f80538f29f124e997da27948c904,} 
Image is up to date for sha256:b360b154b122e3bb16ca12c70517ba0dce04f80538f29f124e997da27948c904
[root@hcm06 ~]# 
```

The error regarding "server misbehaving" for the DNS lookup is expected: the DNS server is `dnsmasq`, which returns "server error" responses for unknown DNS names. Unfortunately, the activity on the mirror isn't evident in the debug logs: it took me a while to figure out that the DNS lookup was happening as fallback because the image fetch via the mirror failed, as opposed to the mirror not being used at all. Here's output from `crictl image ls` for reference.

```
IMAGE                                        TAG                          IMAGE ID            SIZE
docker.io/library/hello-world                latest                       d1165f2212346       6.62kB
docker.io/rancher/hardened-calico            v3.19.1-build20210611        4180617675d1c       185MB
docker.io/rancher/hardened-etcd              v3.4.13-k3s1-build20210223   271c0a695260e       47.8MB
docker.io/rancher/hardened-flannel           v0.14.0-build20210611        a1f72b883ae8e       97.2MB
docker.io/rancher/hardened-kube-proxy        v1.21.1-build20210520        d47665fc14cdb       86MB
docker.io/rancher/hardened-kubernetes        v1.21.2-rke2r1               d70560d45c77b       180MB
docker.io/rancher/nginx-ingress-controller   nginx-0.46.0-rancher1        97417ad4192f8       107MB
docker.io/rancher/pause                      3.2                          e004ddc1b078f       300kB
quay.io/strimzi/operator                     latest                       b360b154b122e       248MB
```

The "latest" image is not the same as the 0.24.0 one: `docker` says the image ID for that is 2d4cedc7092c.

The only thing remotely distinctive about the failing case is that its version number starts with a digit, but is not numeric. Some of the other images have zero-point versions, but they all start with a "v" or name (like "nginx"). Could this be some subtle conversion bug resulting from an attempt to treat the tag as a number?

**Comment by brandond** [comment]
What does your registries.yaml look like?

**Comment by TFBW** [comment]
Due to a certain amount of flailing around and trying things, it looks like this:
```
mirrors:
  docker.io:
    endpoint:
      - "http://proxy:5000"
  quay.io:
    endpoint:
      - "http://proxy:5000"
  "*":
    endpoint:
      - "http://proxy:5000"
```
I plan to remove the explicit names and just leave the wildcard. Everything has to go through the proxy registry.

**Comment by TFBW** [comment]
Looking at the YAML of the strimzi-cluster-operator, I see a reference to ImagePullPolicy with empty braces. Is that what you wanted to know?

**Comment by brandond** [comment]
Hmm. I am guessing that the registry proxy at proxy:5000 is a pull-through for docker.io? You're not going to be able to pull quay.io images from docker.io. If you look at which tags are available for the strimzi/operator image on docker.io, that probably matches the ones you can pull successfully:

https://hub.docker.com/r/strimzi/operator/tags?page=1&ordering=last_updated

You probably want to set up a separate proxy for quay.io, and configure that as the endpoint for quay.io in your registries.yaml.

**Comment by TFBW** [comment]
The exact behaviour of the docker registry is still a mystery to me, and documentation is thin. I am able to pull the correct image using `docker pull`, and I assumed I was going via the same route, but it's probably going via the generic HTTP proxy. It's hard to keep track of all the places I've set those environment variables, and it's equally hard to keep track of which process is referring to what set. I tried to disable it explicitly for testing purposes, but in retrospect my shell environment probably isn't the one that mattered.

On that note, I find that the rke2-server.service file hauls environment variables in from a few locations, none of which I'm currently using. I'll try putting my HTTP proxy settings in one of those locations and see if that doesn't convince `containerd` to use the general HTTP proxy when falling back. If not, then how does one tell `containerd` about proxies? The failure of `containerd` to recognise my proxy settings was one of the reasons I opted for a pull-through registry in the first place: I found documentation on how to configure that.

Maybe I'll try setting up a quay.io pull-through, but I'm already sour on the concept, given that I had to discover via trial and error that the docker.io one won't work unless you have a dockerhub account. I am really starting to miss the days when you could just save bandwidth using squid. Is there no such thing as a generic caching proxy for container images? As in one that doesn't need to be tied to a specific registry? That's all I want for Christmas, really.

**Comment by brandond** [comment]
The thing you're calling a proxy isn't a HTTP proxy. It's a pull-through caching registry mirror. If you ask it for an image and it doesn't have it, it goes out and gets it from somewhere else, keeps a copy locally, and sends it back to you. By default it goes and gets things from docker.io; if you want it to go get things from somewhere else (say quay.io) you need to tell it to do so.

**Comment by TFBW** [comment]
I'm using the pull-through caching registry as a proxy. My experimental cluster is on an isolated subnet, and there is no ability to pull anything except via a proxy of one sort or another. I previously set up a mainline Rancher cluster, and it worked via my HTTP proxy without difficulty; setting up RKE2 had no such luck, but I found the configuration options for the registry mirror and decided to configure that as a proxy, since the HTTPS-everywhere nature of the process meant my squid proxy had no ability to cache anything. My expectation with regards to the pull-through registry was that it wouldn't be tied to a specific upstream source. I *really* expected there to be such a thing as a generic, caching registry proxy, but apparently not?

In any case, this is not a bug: it's a PEBKAC brought on by the fact that although HTTP proxying is supported, it's *very* hard to find all the places that the environment variables need to be set. For the benefit of possible future searches along the lines of, "how do I make containerd use an HTTP proxy in RKE2," the answer is "the rke2-server/rke2-agent services are responsible for starting containerd, and will provide environment variables to it, so look in their systemd service files for appropriate ways to set them." For my part, I created /etc/default/rke2-server, and I'm no longer dependent on the pull-through as my only source of outside connectivity.

Thank you for your assistance.
